spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/AssignmentAlignmentSupport.scala

```diff
@@ -0,0 +1,193 @@
+              val colName = (namePrefix :+ col.name).mkString(".")
+              throw new AnalysisException(
+                "Updating nested fields is only supported for StructType " +
+                s"but $colName is of type $otherType"
+              )
+          }
+
+        // if there are conflicting updates, throw an exception
+        // there are two illegal scenarios:
+        // - multiple updates to the same column
+        // - updates to a top-level struct and its nested fields (e.g., a.b and a.b.c)
+        case updates if hasExactMatch(updates, col, resolver) =>
+          val conflictingCols = updates.map(u => (namePrefix ++ u.nameParts).mkString("."))
+          throw new AnalysisException(
+            "Updates are in conflict for these columns: " +
+            conflictingCols.distinct.mkString("[", ", ", "]"))
+      }
+    }
+  }
+
+  private def applyStructUpdates(
+      col: NamedExpression,
+      updates: Seq[ColumnUpdate],
+      fields: Seq[StructField],
+      resolver: Resolver,
+      namePrefix: Seq[String]): Expression = {
+
+    // build field expressions
+    val fieldExprs = fields.zipWithIndex.map { case (field, ordinal) =>
+      Alias(GetStructField(col, ordinal, Some(field.name)), field.name)()
+    }
+
+    // recursively apply this method on nested fields
+    val newUpdates = updates.map(u => u.copy(nameParts = u.nameParts.tail))
+    val updatedFieldExprs = applyUpdates(fieldExprs, newUpdates, resolver, namePrefix :+ col.name)
+
+    // construct a new struct with updated field expressions
+    toNamedStruct(fields, updatedFieldExprs)
+  }
+
+  private def toNamedStruct(fields: Seq[StructField], fieldExprs: Seq[Expression]): Expression = {
+    val namedStructExprs = fields.zip(fieldExprs).flatMap { case (field, expr) =>
+      Seq(Literal(field.name), expr)
+    }
+    CreateNamedStruct(namedStructExprs)
+  }
+
+  private def hasExactMatch(
+      updates: Seq[ColumnUpdate],
+      col: NamedExpression,
+      resolver: Resolver): Boolean = {
+
+    updates.exists(assignment => isExactMatch(assignment, col, resolver))
+  }
+
+  private def isExactMatch(
+      update: ColumnUpdate,
+      col: NamedExpression,
+      resolver: Resolver): Boolean = {
+
+    update.nameParts match {
+      case Seq(namePart) if resolver(namePart, col.name) => true
+      case _ => false
+    }
+  }
+
+  protected def castIfNeeded(
+      expr: Expression,
+      dataType: DataType,
+      resolver: Resolver): Expression = expr match {
+    // some types cannot be casted from NullType (e.g. StructType)
+    case Literal(value, NullType) => Literal(value, dataType)
+    case _ =>
+      (expr.dataType, dataType) match {
+        // resolve structs by name if they they have the same number of fields and their names match
+        // e.g., it is ok to set a struct with fields (a, b) as another struct with fields (b, a)
+        // it is invalid to a set a struct with fields (a, d) as another struct with fields (a, b)
```

Comment:
This means that assignments within a struct are always done by name and never by position? In table output resolution, it looks like structs must match the expected order and if writing by name then the names must match. I would probably use the same logic as in [`canWrite`](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataType.scala#L493).

Reply:
I gave this a bit of thought.

We should discuss whether to resolve nested struct fields by name or by position. In particular, what should happen if one sets a struct with fields (b, a) to a struct column with fields (a, b)? Should we match the fields by name or should we match them by position?

I think actions in MERGE operations are resolved by position.

For example, `INSERT (c1, c2) VALUES (s.c2, s.c1)` will be resolved as `c1 = s.c2, c2 = s.c1`. I think we can follow what Spark does for INSERTs and resolve struct fields by position, not by name like this PR does now. 

Not resolving nested fields by name may lead to unexpected results from the user perspective but we have this problem inside INSERT statements already. Spark ignores column names inside INSERTs and adds aliases if needed.

Are we ok ignoring column names inside MERGE too? I guess that would be reasonable and would match the existing behavior for SQL INSERTs. If so, assigning a struct (b, a) as struct (a, b) may be legal but the new value of `b` will be applied to `a` column.

Reply:
> In particular, what should happen if one sets a struct with fields (b, a) to a struct column with fields (a, b)? Should we match the fields by name or should we match them by position?

SQL writes are always by position unless you have `(names...) VALUES (values...)`, but that's not the case for these structs. So I think think the right behavior for SQL is by position.

Because dataframe columns don't have an obvious position, the expectation for users is that the writes happen by name. That's why we added `byName` variants of the logical plans. I think that extends to nested structs as well because nested structs are easy to produce by some conversion from an object. When converting from an object using an `Encoder`, there is no column order guarantee so it is reasonable to assume that columns will be written by name.

Because we don't currently support dataframe merge into, I think that we should move forward with the behavior you've added that matches the insert behavior. We can add the `byName` flag later when we add a dataframe API.

Reply:
Well, we do have `(names...) VALUES (values...)` in the insert action:

```
WHEN NOT MATCHED THEN
  INSERT (c1, c2) VALUES (v1, v2)
```

But I am not sure whether that means nested structs must be resolved by name.